persistent/Database/Persist/Quasi.hs

```diff
@@ -197,18 +461,95 @@ removeSpaces =
     toLine' i = Line i . mapMaybe fromToken
 
     fromToken (Token t) = Just t
+    fromToken (DocComment t) = Just t
     fromToken Spaces{}  = Nothing
 
 -- | Divide lines into blocks and make entity definitions.
 parseLines :: PersistSettings -> [Line] -> [EntityDef]
 parseLines ps lines =
     fixForeignKeysAll $ toEnts lines
   where
-    toEnts (Line indent (name:entattribs) : rest) =
-        let (x, y) = span ((> indent) . lineIndent) rest
-        in  mkEntityDef ps name entattribs x : toEnts y
-    toEnts (Line _ []:rest) = toEnts rest
-    toEnts [] = []
+    toEnts :: [Line] -> [UnboundEntityDef]
+    toEnts =
+        map mk
+        . associateLines
+        . skipEmpty
+    mk :: LinesWithComments -> UnboundEntityDef
+    mk lwc =
+        let Line _ (name :| entAttribs) :| rest = lwcLines lwc
+         in setComments (lwcComments lwc) $ mkEntityDef ps name entAttribs (map (mapLine NEL.toList) rest)
+
+isComment :: Text -> Maybe Text
+isComment xs =
+    T.stripPrefix "-- | " xs
+
+data LinesWithComments = LinesWithComments
+    { lwcLines :: NonEmpty (Line' NonEmpty)
+    , lwcComments :: [Text]
+    } deriving (Eq, Show)
+
+-- TODO: drop this and use <> when 8.2 isn't supported anymore so the
+-- monoid/semigroup nonsense isn't annoying
+appendLwc :: LinesWithComments -> LinesWithComments -> LinesWithComments
+appendLwc a b =
+    LinesWithComments (foldr NEL.cons (lwcLines b) (lwcLines a)) (lwcComments a `mappend` lwcComments b)
+
+newLine :: Line' NonEmpty -> LinesWithComments
+newLine l = LinesWithComments (pure l) []
+
+firstLine :: LinesWithComments -> Line' NonEmpty
+firstLine = NEL.head . lwcLines
+
+consLine :: Line' NonEmpty -> LinesWithComments -> LinesWithComments
+consLine l lwc = lwc { lwcLines = NEL.cons l (lwcLines lwc) }
+
+consComment :: Text -> LinesWithComments -> LinesWithComments
+consComment l lwc = lwc { lwcComments = l : lwcComments lwc }
+
+associateLines :: [Line' NonEmpty] -> [LinesWithComments]
+associateLines lines =
+    foldr combine [] $
+    foldr toLinesWithComments [] lines
```

Comment:
this has two `foldr` steps to reduce complexity, and also because it's genuinely easier to work on. happy for refactoring suggestions.